File: data_structures/queues/priority_queue.cpp
#include<iostream>
#include<bits/stdc++.h>

using namespace std;


/*
assuming min heap
*/
class Heap {
public:
    // O(logn), O(n)
    void heapify_top_down(int from) {
        int left = 2 * from + 1, right = 2 * from + 2;

        int smaller = from;
        if (left < size && heap[smaller] > heap[left]) {
            smaller = left;
        }
        
        if (right < size && heap[smaller] > heap[right]) {
            smaller = right;
        }

        if (smaller != from) {
            swap(heap[from], heap[smaller]);
            heapify_top_down(smaller);
        }
    }

    // O(logn), O(n)
    void heapify_bottom_up(int from) {
        int parent = from & 1 ? (from - 1) / 2 : (from - 2) / 2;
        if (parent >= 0 && heap[from] < heap[parent]) {
            swap(heap[from], heap[parent]);
            heapify_bottom_up(parent);
        }
    }

    // O(logn), O(1)
    void push(int value) {
        ++size;
        heap.push_back(value);
        heapify_bottom_up(size - 1);
    }

    // O(logn), O(1)
    int pop() {
        int top = heap[0];
        swap(heap[0], heap[size - 1]);

        --size;
        heap.pop_back();

        heapify_top_down(0);

        return top;
    }

    // O(1), O(1)
    int top() {
        return heap[0];
    }

    // O(1), O(1)
    bool empty() {
        return size == 0;
    }

    // O(n), O(1)
    void print() {
        for (auto& node : heap) {
            cout << node << " ";
        }
        cout << endl;
    }

private:
    int size = 0;
    vector<int> heap;
};


// int main() {
//     Heap heap;

//     for (int i = 10; i >= 0; --i) {
//         heap.push(i);
//     }
//     heap.print();

//     heap.pop();
//     heap.print();
//     heap.pop();
//     heap.print();

//     heap.push(-1);
//     heap.print();

//     cout << "top: " << heap.top() << endl;

//     while (!heap.empty()) {
//         cout << heap.pop() << " ";
//     }
//     cout << endl;

//     return 0;
// }

/*
                        0
            1                       5
    4               2       9               6
10      7       8       3


                        2
            8                       5
    4               3       9               6
10      7


                        2
            4                       5
    7               3      9                6
10      8


                        -1
            2                       5
    7               4       9               6
10      8       3
*/

File: data_structures/queues/priority_queue.py
'''
assuming a min heap
'''

# O(logn), O(1)
# heapifies starting from the given node till the bottom
def heapify(heap, size, current):
    largest = current
    left, right = 2 * current + 1, 2 * current + 2

    if left < size and heap[left] > heap[largest]:
        largest = left
    
    if right < size and heap[right] > heap[largest]:
        largest = right

    if largest != current:
        heap[current], heap[largest] = heap[largest], heap[current]
        heapify(heap, size, largest)


# O(logn), O(1)
# heapifies starting from the given node till the top
# using this for heappush is faster as we can append an item to the end and heapify till the top
def heapify_reverse(heap, current):
    while True:
        if current & 1:
            parent = (current - 1) // 2
        else:
            parent = (current - 2) // 2

        if parent < 0 or heap[current] <= heap[parent]:
            break

        heap[current], heap[parent] = heap[parent], heap[current]
        current = parent


# O(logn), O(1)
# append an element to the heap
def heappush(heap, item):
    heap.append(item)
    heapify_reverse(heap, len(heap) - 1)


# O(logn), O(1)
# pop the top of the heap
def heappop(heap):
    heap[0], heap[len(heap) - 1] = heap[len(heap) - 1], heap[0]
    popped = heap.pop()
    heapify(heap, len(heap), 0)
    return popped


# O(nlogn), O(1)
# reorder to array to a heap in-place such that for any index i, (2i + 1) and (2i + 2) are the left and the right children respectively
def build(arr):
    for i in range((len(arr) // 2) - 1, -1, -1):
        heapify(arr, len(arr), i)


if __name__ == "__main__":
    arr = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3]
    buildarr = arr
    build(buildarr)
    pusharr = []
    for a in arr:
        heappush(pusharr, a)
    
    assert buildarr == pusharr

    print(heappop(pusharr))
    print(pusharr)
    print(heappop(pusharr))
    print(pusharr)
    print(heappop(pusharr))
    print(pusharr)
    print(heappop(pusharr))
    print(pusharr)
